pandas/io/_plugin_loader.py

```diff
@@ -0,0 +1,154 @@
+"""
+Load I/O plugins from third-party libraries into the pandas namespace.
+
+Third-party libraries defining I/O plugins register an entrypoint in
+the `dataframe.io` group. For example:
+
+```
+[project.entry-points."dataframe.io"]
+repr = "pandas_repr:ReprDataFrameIO"
+```
+
+The class `ReprDataFrameIO` will implement readers and writers in at
+least one of different exchange formats supported by the protocol.
+For now a pandas DataFrame or a PyArrow table, in the future probably
+nanoarrow, or a Python wrapper to the Arrow Rust implementations.
+For example:
+
+```python
+class FancyFormatDataFrameIO:
+    @staticmethod
+    def pandas_reader(self, fname):
+        with open(fname) as f:
+            return eval(f.read())
+
+    def pandas_writer(self, fname, mode='w'):
+        with open(fname, mode) as f:
+            f.write(repr(self))
+```
+
+If the I/O plugin implements a reader or writer supported by pandas,
+pandas will create a wrapper function or method to call the reader or
+writer from the pandas standard I/O namespaces. For example, for the
+entrypoint above with name `repr` and methods `pandas_reader` and
+`pandas_writer` pandas will create the next functions and methods:
+
+- `pandas.read_repr(...)`
+- `pandas.Series.to_repr(...)`
+- `pandas.DataFrame.to_repr(...)`
+
+The reader wrappers validates that the returned object is a pandas
+DataFrame when the exchange format is `pandas`, and will convert the
+other supported objects (e.g. a PyArrow Table) to a pandas DataFrame,
+so the result of `pandas.read_repr` is a pandas DataFrame, as the
+user would expect.
+
+If more than one reader or writer with the same name is loaded, pandas
+raises an exception.
+"""
+import functools
+from importlib.metadata import entry_points
+
+import pandas as pd
+
+supported_exchange_formats = ["pandas"]
+
+try:
+    import pyarrow as pa
+except ImportError:
+    pa = None
+else:
+    supported_exchange_formats.append("pyarrow")
+
+
+def _create_reader_function(io_plugin, exchange_format):
+    """
+    Create and return a wrapper function for the original I/O reader.
+
+    We can't directly call the original reader implemented in
+    the connector, since we want to make sure that the returned value
+    of `read_<whatever>` is a pandas DataFrame, so we need to validate
+    it and possibly cast it.
+    """
+    original_reader = getattr(io_plugin, f"{exchange_format}_reader")
+
+    # TODO: Create this function dynamically so the resulting signature contains
+    # the original parameters and not `*args` and `**kwargs`
+    @functools.wraps(original_reader)
+    def reader_wrapper(*args, **kwargs):
+        result = original_reader(*args, **kwargs)
+        if exchange_format == "pyarrow":
```

Comment:
Would be be more generic to convert `result` if it implements the interchange protocol instead? Such that 

1. If `original_reader` uses/returns Arrow objects, `original_reader` can also customize how the Arrow objects are converted to pd.DataFrame via `*args, **kwargs` 
2. Maybe a little more future proof to Arrow as a format?

Reply:
Yes, that's a very good point @mroeschke. Both PyArrow Table and pandas DataFrame support it, so using the interchange protocol in the pandas side of this connectors API seems to solve the problem in a much simpler way.

I don't think there shouldn't be performance implications on using it, but maybe worth to check if data is not being copied with the current implementations of the interchange protocol.

I updated this PR and the lance-reader repo to use what you suggest, please let me know if this is what you had in mind.